lib/pos_consensus.go

```diff
@@ -507,8 +507,28 @@ func (fc *FastHotStuffConsensus) HandleLocalTimeoutEvent(event *consensus.FastHo
 func (fc *FastHotStuffConsensus) HandleValidatorTimeout(pp *Peer, msg *MsgDeSoValidatorTimeout) error {
 	glog.Infof("FastHotStuffConsensus.HandleLocalTimeoutEvent: Received timeout msg: %s", msg.ToString())
 
-	// No need to hold a lock on the consensus because this function is a pass-through
-	// for the FastHotStuffEventLoop which guarantees thread-safety for its callers.
+	// Hold a read lock on the consensus. We will need to read the blockchain data structure.
+	fc.lock.RLock()
+	defer fc.lock.RUnlock()
+
+	if !fc.fastHotStuffEventLoop.IsRunning() {
+		return errors.Errorf("FastHotStuffConsensus.HandleValidatorTimeout: FastHotStuffEventLoop is not running")
+	}
+
+	// Hold the blockchain's read lock so that the chain cannot be mutated underneath us.
+	// In practice, this is a no-op, but it guarantees thread-safety in the event that other
+	// parts of the codebase change.
+	fc.blockchain.ChainLock.Lock()
+	defer fc.blockchain.ChainLock.Unlock()
```

Comment:
Should we release `fc.blockchain.Chainlock` and `fc.lock` before we call `fc.fastHotStuffEventLoop.ProcessValidatorTimeout`? My thinking is that we only need to hold the lock while we are checking the highQC and that `fc.fastHotStuffEventLoop.ProcessValidatorTimeout` should handle locks on its own?

Reply:
Sure. We're find to re-jig the ordering of the locks so they're held for the minimum amount of time

Reply:
We need to guarantee that the handlers in this file are single threaded, which is why we hold fc.lock at the beginning. We could tune things to be more efficient by reordering the locks for less time, but this comes at the cost of deadlock risk. IMO it's not worth optimizing right now.

In this case, it's a great idea to hold and release the fc.blockchain only while performing the blockIndex lookup. I'll update to that 👍